file_writer_control/InThreadStatusTracker.py

```diff
@@ -0,0 +1,212 @@
+from file_writer_control.CommandStatus import CommandStatus, CommandState
+from file_writer_control.WorkerStatus import WorkerState, WorkerStatus
+from file_writer_control.StateExtractor import (
+    extract_worker_state_from_status,
+    extract_state_from_command_answer,
+    extract_job_state_from_answer,
+)
+import json
+from streaming_data_types.action_response_answ import Response
+from typing import Dict
+
+STATUS_MESSAGE_TIMEOUT = timedelta(seconds=5)
+JOB_STATUS_TIMEOUT = timedelta(seconds=5)
+COMMAND_STATUS_TIMEOUT = timedelta(seconds=20)
+
+
+class InThreadStatusTracker:
+    """
+    Implements de-coding of flatbuffer messages and sends updates of worker, job and command state/status back to the
+    "main"-thread if there has been changes.
+    """
+
+    def __init__(self, status_queue: Queue):
+        """
+        Constructor.
+        :param status_queue: The output queue to which state/status updates are pushed.
+        """
+        self.queue = status_queue
+        self.known_workers: Dict[str, WorkerStatus] = {}
+        self.known_jobs: Dict[str, JobStatus] = {}
+        self.known_commands: Dict[str, CommandStatus] = {}
+
+    def process_message(self, message: bytes):
+        """
+        Process a binary message.
+        :param message: The binary message to be processed.
+        """
+        current_schema = _get_schema(message).encode("utf-8")
+        update_time = datetime.now()
+        if current_schema == ANSW_IDENTIFIER:
+            self.process_answer(deserialise_answer(message))
+        elif current_schema == STAT_IDENTIFIER:
+            self.process_status(deserialise_status(message))
+        elif current_schema == STOP_TIME_IDENTIFIER:
+            self.process_set_stop_time(deserialise_stop_time(message))
+        elif current_schema == START_IDENTIFIER:
+            self.process_start(deserialise_start(message))
+        elif current_schema == STOPPED_IDENTIFIER:
+            self.process_stopped(deserialise_stopped(message))
+        else:
+            self.process_unknown(message)
+        self.send_status_if_updated(update_time)
+
+    def process_unknown(self, message: bytes):
+        pass  # Do nothing for now
+
+    def send_status_if_updated(self, limit_time: datetime):
+        """
+        Sends status updates of workers, jobs and commands (to the status queue) if there has been any updates on or
+        after the limit_time.
+        :param limit_time: The cut-off time for deciding which updates should be sent to the status queue.
+        """
+        for worker in self.known_workers.values():
+            if worker.last_update >= limit_time:
+                self.queue.put(worker)
+        for job in self.known_jobs.values():
+            if job.last_update >= limit_time:
+                self.queue.put(job)
+        for command in self.known_commands.values():
+            if command.last_update >= limit_time:
+                self.queue.put(command)
```

Comment:
These three for loops can be made into one via `zip`. That is we may write `for (x, y, z) in zip(k, l, m):`. Nothing to worry about if you find explicit loops more readable.

Reply:
I do not actually much care for `zip` so unless you insist, I would like to leave the code as is.

Reply:
No problem at all.